ChronoVisor/src/KeeperRegistry.cpp

```diff
@@ -505,7 +823,237 @@ int KeeperRegistry::notifyKeepersOfStoryRecordingStop(std::vector<KeeperIdCard>
+/////////////////
+
+int KeeperRegistry::unregisterGrapherProcess(GrapherIdCard const& grapher_id_card)
+{
+    std::lock_guard<std::mutex> lock(registryLock);
+    //check again after the lock is acquired
+    if(is_shutting_down()) { return chronolog::CL_ERR_UNKNOWN; }
+
+    auto group_iter = recordingGroups.find(grapher_id_card.getGroupId());
+    if(group_iter == recordingGroups.end()) { return chronolog::CL_SUCCESS; }
+
+    RecordingGroup& recording_group = ((*group_iter).second);
+
+    std::stringstream id_string;
+    id_string << grapher_id_card;
+    if(recording_group.grapherProcess != nullptr && recording_group.grapherProcess->active)
+    {
+        // start delayed destruction for the lingering Adminclient to be safe...
+        // to prevent the case of deleting the keeperAdminClient while it might be waiting for rpc response on the
+        // other thread
+
+        std::time_t delayedExitTime = std::chrono::high_resolution_clock::to_time_t(
+                std::chrono::high_resolution_clock::now() + std::chrono::seconds(delayedDataAdminExitSeconds));
+        LOG_INFO("[KeeperRegistry] grapher {} starting delayedExit for grapher {} delayedExitTime={}", id_string.str(),
+                 std::ctime(&delayedExitTime));
+
+        recording_group.startDelayedGrapherExit(*(recording_group.grapherProcess), delayedExitTime);
+    }
+
+    // now that we are still holding registryLock
+    // update registryState if needed
+    LOG_INFO("[KeeperRegistry] RecordingGroup {} has no grappher and  {} keepers", recording_group.groupId,
+             recording_group.keeperProcesses.size());
+
+    // we've just unregistered the grapher so the group can't perform recording duties any longer
+    {
+        activeGroups.erase(std::remove(activeGroups.begin(), activeGroups.end(), &recording_group));
+        if(activeGroups.size() > 0)
+        {//reset the group distribution
+            size_t new_seed = std::chrono::high_resolution_clock::to_time_t(std::chrono::high_resolution_clock::now());
+            mt_random.seed(new_seed);//re-seed the mersene_twister_generator
+            group_id_distribution =
+                    std::uniform_int_distribution<size_t>(0, activeGroups.size() - 1);//reset the distribution range
+        }
+    }
+
+    LOG_INFO("[KeeperRegistry]  has {} RecordingGroups ; {} activeGroups", recordingGroups.size(), activeGroups.size());
+    // update registryState in case this was the last active recordingGroup
+    if(activeGroups.size() == 0) { registryState = INITIALIZED; }
+
+    return chronolog::CL_SUCCESS;
+}
 
 }//namespace chronolog
 
 
+///////////////
+
+void chl::RecordingGroup::startDelayedGrapherExit(chl::GrapherProcessEntry& grapher_process,
+                                                  std::time_t delayedExitTime)
+{
+    grapher_process.active = false;
+
+    LOG_DEBUG("[KeeperRegistry] recording_group {} starts delayedExit for {}", groupId,
+              grapher_process.idCardString.str());
```

Comment:
do you mean `grapher_process.idCardString.c_str()` or just `grapher_process.idCardString`? same for line 1022, 1039, 1054.

Reply:
fixed in the latest commit